### PR TITLE
Setup docdash for the project

### LINKED ISSUE #83: Provide API documentation
http://clenemt.github.io/docdash/

### PR DESCRIPTION
**Pull request by tawfiek**
Fixes #83 
It looks pretty well and it will make starting with the codebase much easier, but  I think we need to configure it to fit our needs.
@bwindels 
Screenshot from 2021-03-28 13-59-51

### CODE REVIEW

**Comment by bwindels** [comment]
Hey @tawfiek, great work! It'll be really useful to have some API docs.
I checkout out your branch, did `yarn install` and `yarn generate-docs` and all I get as output is this though:
```
$ node_modules/.bin/jsdoc -c jsdoc.json
There are no input files to process.
```
With the --debug option:
```
$ node_modules/.bin/jsdoc --debug -c jsdoc.json
DEBUG: JSDoc 3.6.6 (Sun, 20 Sep 2020 02:25:14 GMT)
DEBUG: Environment info: {"env":{"conf":{"plugins":["plugins/markdown"],"recurseDepth":10,"source":{"includePattern":"\\.js$","excludePattern":"(node_modules/|docs)","include":"src/main.js"},"sourceType":"module","tags":{"allowUnknownTags":false,"dictionaries":["jsdoc","closure"]},"templates":{"monospaceLinks":false,"cleverLinks":false},"opts":{"encoding":"utf8","destination":"docs/","recurse":true,"verbose":true},"docdash":{"static":true,"collapse":true,"wrap":true}},"opts":{"_":[],"debug":true,"configure":"jsdoc.json","encoding":"utf8","destination":"docs/","recurse":true,"verbose":true}}}
There are no input files to process.
```

File: package.json [hunk 1]
```diff
@@ -45,15 +46,17 @@
...
   },
   "dependencies": {
-    "rollup": "^2.26.4",
     "@rollup/plugin-commonjs": "^15.0.0",
     "@rollup/plugin-node-resolve": "^9.0.0",
     "aes-js": "^3.1.2",
     "another-json": "^0.2.0",
     "base64-arraybuffer": "^0.2.0",
     "bs58": "^4.0.1",
+    "docdash": "^1.2.0",
```
Comment (bwindels) [inline package.json]: These should be added as `devDependencies` rather than regular dependencies.

**Review by bwindels — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (tawfiek) [reply package.json]: sure 👍🏽

**Comment by tawfiek** [comment]
> Hey @tawfiek, great work! It'll be really useful to have some API docs.
> I checkout out your branch, did `yarn install` and `yarn generate-docs` and all I get as output is this though:
> 
> ```
> $ node_modules/.bin/jsdoc -c jsdoc.json
> There are no input files to process.
> ```
> 
> With the --debug option:
> 
> ```
> $ node_modules/.bin/jsdoc --debug -c jsdoc.json
> DEBUG: JSDoc 3.6.6 (Sun, 20 Sep 2020 02:25:14 GMT)
> DEBUG: Environment info: {"env":{"conf":{"plugins":["plugins/markdown"],"recurseDepth":10,"source":{"includePattern":"\\.js$","excludePattern":"(node_modules/|docs)","include":"src/main.js"},"sourceType":"module","tags":{"allowUnknownTags":false,"dictionaries":["jsdoc","closure"]},"templates":{"monospaceLinks":false,"cleverLinks":false},"opts":{"encoding":"utf8","destination":"docs/","recurse":true,"verbose":true},"docdash":{"static":true,"collapse":true,"wrap":true}},"opts":{"_":[],"debug":true,"configure":"jsdoc.json","encoding":"utf8","destination":"docs/","recurse":true,"verbose":true}}}
> There are no input files to process.
> ```

I'll remove `node_module` and try to regenerate the issue

**Comment by tawfiek** [comment]
> I checkout out your branch, did `yarn install` and `yarn generate-docs` and all I get as output is this though:

I couldn't regenerate the issue.
However according to this [comment in jsdoc repo](https://github.com/jsdoc/jsdoc/issues/1328#issuecomment-282143743) I  removed `excludePattern` from the config file.

**Comment by tawfiek** [comment]
any updates here ?! 
@bwindels

**Comment by bwindels** [comment]
No updates, I still have to retry with your `excludePattern` changes.

**Comment by bwindels** [comment]
Right, it seems to work now in that I get the same output as in your screenshot. Before landing this, it would be good to figure out how to get it to output something more useful, as it currently only seems to pick up a few classes and methods. Is it only picking up declarations with a jsdoc comment perhaps? Is there a way to also have it generate API docs for classes, methods, ... without jsdoc comments?

**Review by bwindels — CHANGES_REQUESTED** [review]
See comment ^

**Comment by tawfiek** [comment]
Yes, It dose pickup declarations with a jsdoc comments only.
In fact , docdash is nothing but a documentation template theme for JSDoc.
As in `react-sdk` , I think we need to put some efforts to make the code more documented, and strict our contributions rules.

**Comment by bwindels** [comment]
> Yes, It dose pickup declarations with a jsdoc comments only.
> In fact , docdash is nothing but a documentation template theme for JSDoc.
> As in `react-sdk` , I think we need to put some efforts to make the code more documented, and strict our contributions rules.

Is there any way for jsdoc or any other documentation tool to also generate API docs for code missing jsdoc comments?

**Comment by tawfiek** [comment]
I don't think so.
As javaScript is a loosely typed language, so how can any tool obtain what should the input or output types of the function objects? 
Are this function represent a class constructor or not ?! 
The JsDocs comments here is the tool who's add this sense to the tools like docdash, Minami, doxdox or even to your text editor, to help you auto complete while writing the code or adding a signature to the functions.

For VSCode for example 
Screenshot from 2021-05-03 19-35-00
This image before adding JSDoc comment 

Screenshot from 2021-05-03 19-40-34

And this the function signature after adding the comments.

The VSCode here just reading the comment and renders it in the nice form to give you some information about this function, and that's exactly what other tools does.

**Comment by bwindels** [comment]
> I don't think so.
> As javaScript is a loosely typed language, so how can any tool obtain what should the input or output types of the function objects?

I thought given that it is parsing the code anyway, jsdoc may still generate documentation for code although it does not know the type (e.g. say there is a property `x` of type unknown on class `Y`), but guess not :(

If having this level of API documentation requires jsdoc comments everywhere, it becomes a bit less interesting to have, as that would basically be the work of a typescript conversion and I think I'd then rather count that as another reason to go towards typescript (based on jsdoc comments or actually full typescript) and use something like `tsdoc` perhaps?

Thank you so much for investigating this though, much appreciated!

**Comment by tawfiek** [comment]
Going toward typescript is a good idea anyway.

**Pull request closed without merging**